eth-withdrawal/src/main/kotlin/com/d3/eth/withdrawal/withdrawalservice/WithdrawalServiceImpl.kt

```diff
@@ -75,40 +87,95 @@ class WithdrawalServiceImpl(
             }
     }
 
+    /**
+     * Performs rollback. Return all transferred assets.
+     */
     override fun returnIrohaAssets(event: WithdrawalServiceOutputEvent): Result<Unit, Exception> {
+        logger.info("Withdrawal rollback initiated for Iroha tx ${event}")
+        return getTransfers(event)
+            .map { transfers ->
+                transfers.filter { transferAsset ->
+                    transferAsset.destAccountId == credential.accountId
+                }.map { transfer ->
+                    TransferData(
+                        transfer.srcAccountId,
+                        transfer.assetId,
+                        transfer.amount,
+                        "withdrawal rollback " + transfer.description
+                    )
+                }
+            }.map { transferData ->
+                var transactionBuilder = Transaction
+                    .builder(irohaConsumer.creator, System.currentTimeMillis())
+                transferData.forEach { transfer ->
+                    transactionBuilder = transactionBuilder.transferAsset(
+                        credential.accountId,
+                        transfer.dstAccountId,
+                        transfer.assetId,
+                        transfer.description,
+                        transfer.amount
+                    )
+                }
+                transactionBuilder.build()
+            }.map { transaction ->
+                irohaConsumer.send(transaction)
+            }.map { hash ->
+                logger.info("Successfully sent rollback transaction to Iroha, hash: $hash")
+            }
+    }
+
+    /**
+     * Finalize withdrawal:
+     * 1) Subtract asset
+     * 2) send fee to billing account
+     */
+    override fun finalizeWithdrawal(event: WithdrawalServiceOutputEvent): Result<String, Exception> =
+        getTransfers(event)
+            .map { transfers ->
+                val transfer = transfers.first { it.description != FEE_DESCRIPTION }
+                val fees = transfers.filter { it.description == FEE_DESCRIPTION }
```

Comment:
when is it possible to have one transfer (withdrawal) and many fees? they should be `many` both imo